pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py

```diff
@@ -261,41 +273,124 @@ def zombie_all_players(self, knight_list, archer_list, run):
+                upper_y_bound = min(max_y, const.SCREEN_HEIGHT)
+                lower_x_bound = max(min_x, 0)
+                upper_x_bound = min(max_x, const.SCREEN_WIDTH)
+                startx = lower_x_bound - min_x
+                starty = lower_y_bound - min_y
+                endx = 512 + upper_x_bound - max_x
+                endy = 512 + upper_y_bound - max_y
+                cropped = np.zeros_like(self.observation_spaces[agent].low)
+                cropped[startx:endx, starty:endy, :] = screen[
+                    lower_x_bound:upper_x_bound, lower_y_bound:upper_y_bound, :
+                ]
+
+            return np.swapaxes(cropped, 1, 0)
 
-        if not agent_obj.alive:
-            cropped = np.zeros((512, 512, 3), dtype=np.uint8)
         else:
-            min_x = agent_position[0] - 256
-            max_x = agent_position[0] + 256
-            min_y = agent_position[1] - 256
-            max_y = agent_position[1] + 256
-            lower_y_bound = max(min_y, 0)
-            upper_y_bound = min(max_y, const.SCREEN_HEIGHT)
-            lower_x_bound = max(min_x, 0)
-            upper_x_bound = min(max_x, const.SCREEN_WIDTH)
-            startx = lower_x_bound - min_x
-            starty = lower_y_bound - min_y
-            endx = 512 + upper_x_bound - max_x
-            endy = 512 + upper_y_bound - max_y
-            cropped = np.zeros_like(self.observation_spaces[agent].low)
-            cropped[startx:endx, starty:endy, :] = screen[
-                lower_x_bound:upper_x_bound, lower_y_bound:upper_y_bound, :
-            ]
-
-        return np.swapaxes(cropped, 1, 0)
+            # get the agent
+            i = self.agent_name_mapping[agent]
+            agent = self.agent_list[i]
+            agent_state = agent.vector_state
+
+            # get the agent position
+            agent_pos = np.expand_dims(agent_state[:2], axis=0)
+            agent_ang = np.expand_dims(agent_state[2:], axis=0)
+
+            # get vector state of everything
+            state = self.get_vector_state()
+            all_pos = state[:, :2]
+            all_ang = state[:, 2:]
+
+            # compute the angles that everything is facing
+            all_ang = np.arctan2(all_ang[:, 0], all_ang[:, 1])
+            agent_ang = np.arctan2(agent_ang[:, 0], agent_ang[:, 1])
+
+            # get relative angles that everything is facing relative
+            # to the current agent
+            rel_ang = all_ang - agent_ang
+
+            # get rotation matrix
+            c, s = np.cos(rel_ang), np.sin(rel_ang)
+            rot_mat = np.array([[c, -s], [s, c]])
+            rot_mat = np.transpose(rot_mat, (2, 0, 1))
+
+            # unit vector
+            unit_vec = np.array([0, -1])
+
+            # get rotated relative unit vectors
+            rel_ang = rot_mat @ unit_vec
+
+            # get relative positions scaled to screen size
+            rel_pos = all_pos - agent_pos
+            rel_pos = rel_pos / const.SCREEN_DIAG
```

Comment:
This magnitude normalization should also happen when someone calls `state()`

Reply:
Fixed.